File: ANN(Python)/romberg.c
#include <stdio.h>
#include <math.h>

#define ordemEerro 12
// 4 pois a primeira coluna deve ter 'error_order / 2' elementos
#define numElemsFirstCol 6

double trapz(double (*f)(double), double a, double b, int n){
    double soma = 0;
    double h = (double)(b - a) / (double)n;
    for(int i = 1; i < n; i++){
        double xi = a + i * h;
        soma += f(xi);
    }
    soma *= 2;
    soma += f(a);
    soma += f(b);
    soma *= (0.5 * h);
    return soma;
}


void romberg(double array[], int error_order){

    int numCols = error_order / 2 - 1;
    for(int i = 0; i < numCols; i++){
        for(int j = 0; j < numCols; j++){
            double numer = pow(2, (i + 1) * 2) * array[j + 1] - array[j];
            double denom = pow(2, (i + 1) * 2) - 1;
            array[j] = numer / denom;
        }
    }
    printf("Aprox O(h^%d) = %.16f\n", error_order, array[0]);
}


double f(double x){
    return exp(-x*x);
}

int main(){

    //exemplo
    //aprox a integral de exp(-x*x), de 0 a 1

    double a = 0;
    double b = 1;
    double h = 0.25;
    int n = (int)((b - a) / h);


    printf("%.16f\n", (double)n);

    double coluna_F1[numElemsFirstCol] = {};
    for(int i = 0; i < numElemsFirstCol; i++){
        coluna_F1[i] = trapz(f, a, b, (i + 1) * n);
    }
    romberg(coluna_F1, ordemEerro);

}

File: ANN(Python)/cods_p2/romberg.py
import math

#define ordemEerro 4
# 4 pois a primeira coluna deve ter 'error_order / 2' elementos
#define numElemsFirstCol 2

def trapz(f, a, b, n):
    soma = 0
    h = (b - a) / n
    for i in range(1, n):
        soma += f(a + i * h)
    
    soma *= 2
    soma += f(a)
    soma += f(b)
    soma *= (0.5 * h)
    return soma



def romberg(array, error_order):

    numCols = error_order // 2 - 1
    for i in range(numCols - 1):
        for j in range(numCols - 1):
            numer = 2**((i + 1) * 2) * array[j + 1] - array[j]
            denom = 2**((i + 1) * 2) - 1
            array[j] = numer / denom
    print(f'Aprox O(h^{error_order}) = {array[0]}')


def f(x):
  g = 9.81
  m = 65.73
  cd = 0.48
  return math.sqrt((g*m) / cd)*math.tanh(math.sqrt((g*cd) / m) * x)

def f2(x):
    return math.sqrt(1+x**2)

def f3(x):
    return math.cos(-x**2/3)

def f4(x):
    return (x+1/x)**2

def f5(x):
    return math.exp(-x**2)



if __name__ == '__main__':
    #exemplo
    #aprox a integral de exp(-x*x), de 0 a 1


  
    a = [0]
    b = [7.71]
    order = [8]
    h = [7.71/10]


    tam = len(a)
    for k in range(tam):
        coluna_F1 = []
        numElemsFirstCol = order[k] // 2
        n = (int)((b[k] - a[k]) // h[k])
        for i in range(numElemsFirstCol):
            coluna_F1.append(trapz(f, a[k], b[k], 2**i * n))
        romberg(coluna_F1, order[k])